### DISCUSSION TITLE
Looking for a way to turn off model editing on production environments

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#13588] [Archive]
Hi all!

We are using Directus as a backend for a large project with many deployments. Last week we had a case where the models on a server where conflicting with our snapshot schema, and it caused some heavy merge issues. In the end we had to recreate that environment from scratch.

We're looking for a way to disallow the editing of models per environment, so we can only update them using snapshots. I imagine there must be an environment setting somewhere but can't find it in the docs. Could you help me up to speed? Something like `ALLOW_DATA_MODEL_EDITING=false`

If there is no matching functionality for this I'd like to make a feature request for this.

**Comment by br41nslug** [comment]
There is currently no such toggle available in the core. However it is possible to do this with a simple hook extension blocking all edits to the data-model related system collections.
```
export default ({ filter }, { exceptions }) => {
    const { ForbiddenException } = exceptions;
    filter('collections.create', () => { throw new ForbiddenException() });
    filter('collections.update', () => { throw new ForbiddenException() });
    filter('collections.delete', () => { throw new ForbiddenException() });
    filter('fields.create', () => { throw new ForbiddenException() });
    filter('fields.update', () => { throw new ForbiddenException() });
    filter('fields.delete', () => { throw new ForbiddenException() });
}
```

**Comment by ghost** [comment]
Thanks a lot. For anyone coming across this in the future I added this file in `/extensions/hooks/models/index.ts`

```
import { HookConfig } from '@directus/shared/types';

const Hook: HookConfig = ({ filter }, { exceptions }) => {
  const { ForbiddenException } = exceptions;

  /** Disallow model editing on production environments */
  if (process.env.NODE_ENV !== 'development') {
    filter('collections.create', () => {
      throw new ForbiddenException('Model editing is unavailable');
    });
    filter('collections.update', () => {
      throw new ForbiddenException('Model editing is unavailable');
    });
    filter('collections.delete', () => {
      throw new ForbiddenException('Model editing is unavailable');
    });
    filter('fields.create', () => {
      throw new ForbiddenException('Model editing is unavailable');
    });
    filter('fields.update', () => {
      throw new ForbiddenException('Model editing is unavailable');
    });
    filter('fields.delete', () => {
      throw new ForbiddenException('Model editing is unavailable');
    });
  }
};

export default Hook;

```

This ensures Devs can edit locally but it's not allowed on any servers (as long as you properly configure them)
